`terraform_empty_list_equality`: improve expression detection

bmbferreira opened the pull request:
Fixes partially https://github.com/terraform-linters/tflint/issues/1424 to be able to also detect empty lists on more complex conditional expressions such as `var.enabled && var.proxy_subnets != [] ? 1 : 0`.

As @bendrucker referred [here](https://github.com/terraform-linters/tflint/issues/1424#issuecomment-1165980739), the evaluation of complex attribute values is a deeper issue (https://github.com/terraform-linters/tflint/issues/1257) and this PR does not fix that part.

bendrucker commented on rules/terraformrules/terraform_empty_list_equality.go:
```diff
@@ -55,27 +55,33 @@ func (r *TerraformEmptyListEqualityRule) Check(runner *tflint.Runner) error {
 	}
 }
+
+func emitIssue(exprRange hcl.Range, runner *tflint.Runner, r *TerraformEmptyListEqualityRule) {
```
This is not a suitable function name, note that Go is package scoped

bmbferreira replied:
renamed it to `emitEmptyListEqualityIssues`. Now it also receives a map (actually, a Set) with all the issues to emit.

bendrucker commented on rules/terraformrules/terraform_empty_list_equality_test.go:
```diff
@@ -41,6 +41,34 @@ resource "aws_db_instance" "mysql" {
+			Name: "comparing with [] is not recommended (mixed with other conditions)",
+			Content: `
+resource "aws_db_instance" "mysql" {
+	count = true == true || false != true && (false == false || [] == []) ? 1 : 0
```
This test is unclear and seems to be trying to do to much. The goal here should be to have unit tests where the case being tested is clear and unique. That likely means more tests to achieve the same coverage, but the ability for a reader to understand the test's intent is paramount.

bmbferreira replied:
separated this test in two (one to check parentheses and another one to check multiple conditions that check for empty lists)

bendrucker commented on rules/terraformrules/terraform_empty_list_equality_test.go:
```diff
@@ -41,6 +41,34 @@ resource "aws_db_instance" "mysql" {
+	instance_class = "m4.2xlarge"
+}`,
+			Expected: tflint.Issues{
+				{
```
This doesn't seem like a correct result. The expression that is invalid is the comparison, with both left, right, and a comparator. In the case of `[] == []` that is only one invalid expression and should be one issue. Emitting duplicate issues should be a hint that something's off.

bmbferreira replied:
Just used a map to store the ranges and to avoid emitting multiple issues with the same range. Now it will just report once in case we have `[] == []`

bmbferreira commented:
Hi @bendrucker , thanks for the swift review. 🙇 
Addressed all your comments, please let me know if there's anything else to fix/improve. 

Have a great weekend!

bendrucker commented on rules/terraformrules/terraform_empty_list_equality.go:
```diff
@@ -10,6 +10,7 @@ import (
 
 // TerraformEmptyListEqualityRule checks whether is there a comparison with an empty list
 type TerraformEmptyListEqualityRule struct{}
+type void struct{}
```
Better to just use anonymous empty structs rather than inventing a `void` type like this

bendrucker commented on rules/terraformrules/terraform_empty_list_equality.go:
```diff
@@ -55,27 +56,37 @@ func (r *TerraformEmptyListEqualityRule) Check(runner *tflint.Runner) error {
-				}
-			}
+			var issuesRangeSet = make(map[hcl.Range]void)
+			searchEmptyList(conditionalExpr.Condition, runner, r, conditionalExpr.Range(), issuesRangeSet)
```
Making your helper functions into methods on `*TerraformEmptyListEqualityRule` would eliminate the naming issues

bmbferreira replied:
made it a method but kept the name. Also added comments to all the `*TerraformEmptyListEqualityRule` methods to be more clear on what they're doing

bendrucker commented on rules/terraformrules/terraform_empty_list_equality_test.go:
```diff
@@ -30,6 +30,25 @@ resource "aws_db_instance" "mysql" {
+			},
+		},
+		{
+			Name: "comparing with [] multiple times in the same statement is not recommended",
```
You can tighten up these names to be shorter/clearer for Go test output. For example, `"multiple empty comparisons"`.

bendrucker reviewed (commented):
Nice, getting closer!

bendrucker commented:
Thanks, will be away for a few days, can give this another review in about a week.

bmbferreira commented:
> Thanks, will be away for a few days, can give this another review in about a week.

Ok! Thank you! 🙇

wata727 commented:
While working on another issue, added a change to walk expressions inside expressions in `WalkExpressions` https://github.com/terraform-linters/tflint/pull/1432
Perhaps this issue has also been resolved. Can you confirm it?

wata727 commented:
I checked it again on the latest master branch, but it didn't fix the issue. This fix is still needed.

wata727 commented on rules/terraformrules/terraform_empty_list_equality.go:
```diff
@@ -52,30 +52,43 @@ func (r *TerraformEmptyListEqualityRule) Check(runner *tflint.Runner) error {
+			issuesRangeSet[exprRange] = struct{}{}
+		}
+	}
+}
```
Perhaps this search logic can be rewritten fairly easily by #1432.

bmbferreira replied:
Hi @wata727 ! Thanks for the suggestion and sorry for being silent in the last couple of weeks. Would you mind to ellaborate on your suggestion please? 😕 I'm calling `WalkExpressions` [here](https://github.com/terraform-linters/tflint/pull/1426/files/dd24efe8c505b32e3d15b093295eddeb85d55605#diff-c4c162621b4beba8a29758aed0555eb9d6a2e326ae4b2f537f3fad72592432c4R57) but as you noticed the latest change didn't fix the issue, therefore I think I still need to process the expression recursively here.

wata727 replied:
I think the pattern that this rule essentially wants to detect is `any == []`. In other words, it might be better to check against `BinaryOpExpr` while walking through all the expressions, rather than checking for `ConditionalExpr` first.

Imagine something like:

```go
runner.WalkExpression(func(expr hcl.Expression) error {
  if binaryOpExpr, ok := expr.(*hclsyntax.BinaryOpExpr); ok && binaryOpExpr.Op.Type == cty.Bool {
    if tupleConsExpr, ok := binaryOpExpr.RHS.(*hclsyntax.TupleConsExpr); ok && len(tupleConsExpr.Exprs) == 0 {
      runner.EmitIssue(...)
    }

    if tupleConsExpr, ok := binaryOpExpr.LHS.(*hclsyntax.TupleConsExpr); ok && len(tupleConsExpr.Exprs) == 0 {
      runner.EmitIssue(...)
    }
  }
})
```

What do you think?

bmbferreira replied:
ahh! got it 🙇 Just tried it and it is much better now! I could even remove the map to avoid duplicated ranges 👌 thank you!

bendrucker reviewed (changes requested):
One small style change, would apply myself but I don't seem to have edit access to push to your PR branch

```diff
diff --git a/rules/terraformrules/terraform_empty_list_equality.go b/rules/terraformrules/terraform_empty_list_equality.go
index 25baf37a..83bbfd7a 100644
--- a/rules/terraformrules/terraform_empty_list_equality.go
+++ b/rules/terraformrules/terraform_empty_list_equality.go
@@ -58,17 +58,17 @@ func (r *TerraformEmptyListEqualityRule) checkEmptyList(runner *tflint.Runner) e
 	return runner.WalkExpressions(func(expr hcl.Expression) error {
 		if binaryOpExpr, ok := expr.(*hclsyntax.BinaryOpExpr); ok && binaryOpExpr.Op.Type == cty.Bool {
 			if tupleConsExpr, ok := binaryOpExpr.LHS.(*hclsyntax.TupleConsExpr); ok && len(tupleConsExpr.Exprs) == 0 {
-				r.emitEmptyListEqualityIssue(binaryOpExpr.Range(), runner)
+				r.emitIssue(binaryOpExpr.Range(), runner)
 			} else if tupleConsExpr, ok := binaryOpExpr.RHS.(*hclsyntax.TupleConsExpr); ok && len(tupleConsExpr.Exprs) == 0 {
-				r.emitEmptyListEqualityIssue(binaryOpExpr.Range(), runner)
+				r.emitIssue(binaryOpExpr.Range(), runner)
 			}
 		}
 		return nil
 	})
 }
 
-// emitEmptyListEqualityIssue emits issue for comparison with static empty list
-func (r *TerraformEmptyListEqualityRule) emitEmptyListEqualityIssue(exprRange hcl.Range, runner *tflint.Runner) {
+// emitIssue emits issue for comparison with static empty list
+func (r *TerraformEmptyListEqualityRule) emitIssue(exprRange hcl.Range, runner *tflint.Runner) {
 	runner.EmitIssue(
 		r,
 		"Comparing a collection with an empty list is invalid. To detect an empty collection, check its length.",
```

You can copy the above and pipe to `git apply`.

bmbferreira commented:
> One small style change, would apply myself but I don't seem to have edit access to push to your PR branch
> 
> ```diff
> diff --git a/rules/terraformrules/terraform_empty_list_equality.go b/rules/terraformrules/terraform_empty_list_equality.go
> index 25baf37a..83bbfd7a 100644
> --- a/rules/terraformrules/terraform_empty_list_equality.go
> +++ b/rules/terraformrules/terraform_empty_list_equality.go
> @@ -58,17 +58,17 @@ func (r *TerraformEmptyListEqualityRule) checkEmptyList(runner *tflint.Runner) e
>  	return runner.WalkExpressions(func(expr hcl.Expression) error {
>  		if binaryOpExpr, ok := expr.(*hclsyntax.BinaryOpExpr); ok && binaryOpExpr.Op.Type == cty.Bool {
>  			if tupleConsExpr, ok := binaryOpExpr.LHS.(*hclsyntax.TupleConsExpr); ok && len(tupleConsExpr.Exprs) == 0 {
> -				r.emitEmptyListEqualityIssue(binaryOpExpr.Range(), runner)
> +				r.emitIssue(binaryOpExpr.Range(), runner)
>  			} else if tupleConsExpr, ok := binaryOpExpr.RHS.(*hclsyntax.TupleConsExpr); ok && len(tupleConsExpr.Exprs) == 0 {
> -				r.emitEmptyListEqualityIssue(binaryOpExpr.Range(), runner)
> +				r.emitIssue(binaryOpExpr.Range(), runner)
>  			}
>  		}
>  		return nil
>  	})
>  }
>  
> -// emitEmptyListEqualityIssue emits issue for comparison with static empty list
> -func (r *TerraformEmptyListEqualityRule) emitEmptyListEqualityIssue(exprRange hcl.Range, runner *tflint.Runner) {
> +// emitIssue emits issue for comparison with static empty list
> +func (r *TerraformEmptyListEqualityRule) emitIssue(exprRange hcl.Range, runner *tflint.Runner) {
>  	runner.EmitIssue(
>  		r,
>  		"Comparing a collection with an empty list is invalid. To detect an empty collection, check its length.",
> ```
> 
> You can copy the above and pipe to `git apply`.
> 
> Otherwise LGTM

thanks again for the review @bendrucker. just applied the suggested change!

bendrucker commented:
Getting some 500s right now from GitHub for merge status. I'll revisit this and merge tomorrow.

bmbferreira commented:
> Getting some 500s right now from GitHub for merge status. I'll revisit this and merge tomorrow.

yup, np. Seems they're having some issues: https://www.githubstatus.com/
<img width="990" alt="image" src="https://user-images.githubusercontent.com/626180/183785482-1f32ff7a-696c-462e-9f30-40397999aef1.png">
